distributed/protocol/serialize.py

```diff
@@ -52,11 +52,16 @@ def dask_loads(header, frames):
 
 
 def pickle_dumps(x):
-    return {"serializer": "pickle"}, [pickle.dumps(x)]
+    header = {"serializer": "pickle"}
+    frames = [None]
+    buffer_callback = lambda f: frames.append(memoryview(f))
+    frames[0] = pickle.dumps(x, buffer_callback=buffer_callback)
+    return header, frames
 
 
 def pickle_loads(header, frames):
-    return pickle.loads(b"".join(frames))
+    x, buffers = frames[0], frames[1:]
```

Comment:
My expectation is that `frames[1:]` would be an empty list here. It might not be though. If it's not, that would be good to fix. Whether that has to do with `deserialize_bytes` behavior or not, I'm less sure. Maybe someone else knows?

Reply:
`frames` in `pickle_loads` comes from `deserialize_bytes` for me (take a look at exception stack trace). And for me `frames[1:]` is not empty. That's why I am asking if this is expected or not.

Reply:
Right. So do we have a reproducer? That should help us debug this further.

Reply:
FWIW I've been trying to reproduce this too, but haven't had any luck. Though I believe you that there could be a problem here. It's just tricky to come up with fixes in the dark. So anything you can come up with would help 🙂

Reply:
If you say that Dask is able to split frames, may it happen when large pandas `DataFrames` are serialized? My test is a benchmark which operates on such objects, about 20M of records long and 40 columns wide.

Reply:
Indeed it might. In particular am eyeing [this function]( https://github.com/dask/distributed/blob/fbdb067ed514024aad14dbdc3790bae9344548df/distributed/protocol/utils.py#L22 ). Though I have tried creating fairly large NumPy arrays and haven't managed to trigger that case. Still it seems like that is probably what is happening. Just need to narrow it down a bit more.

Reply:
Maybe something like PR ( https://github.com/dask/distributed/pull/3841 ) will help? Still lack a test case though. So I have no way to confirm whether it actually helps (or ensure we don't accidentally regress).

Reply:
Even when sending large bytes that do get split by `frame_split_size`, I'm unable to reproduce an error. Things are properly reassembled on the other side. Not sure what's going on there.

Reply:
Yeah me neither unfortunately. Happy to dig deeper once we identify a reproducer 🙂

Reply:
A similar issue was raised recently ( https://github.com/dask/distributed/issues/3851 ). Not sure if that is the same. Suspect PR ( https://github.com/dask/distributed/pull/3639 ) fixes this. Would be good if you can try though.

Reply:
This was fixed by PR ( https://github.com/dask/distributed/pull/3639 ).